File: cpp/test/libstmpcttest/qdigest_tests.cpp
#include "custom_number_type.hpp"
#include "minimal_number_type.hpp"
#include <gtest/gtest.h>
#include <iostream>
#include <random>
#include <stmpct/qdigest.hpp>
#include <type_traits>
#include <vector>

#ifndef ARRAYSIZE
#define ARRAYSIZE(x) (sizeof(x) / sizeof(x[0]))
#endif

using namespace std;
using namespace stmpct;

TEST(qdigest_summary, builds_correctly) {
    qdigest_summary summary = qdigest_summary::builder(5)
                                  .insert(0)
                                  .insert(2, 4)
                                  .insert(3, 6)
                                  .insert(4, 1)
                                  .insert(5, 1)
                                  .insert(6, 1)
                                  .insert(7, 1)
                                  .build();
    std::stringstream ss;
    ss << summary;
    ASSERT_EQ(ss.str(), R"EOF((0, 7, 1)
+- (0, 3, 0)
|  +- 
|  `- (2, 3, 0)
|     +- (2, 2, 4)
|     `- (3, 3, 6)
`- (4, 7, 0)
   +- (4, 5, 2)
   `- (6, 7, 2))EOF");
}

TEST(qdigest, example_from_paper) {
    qdigest c(5);
    c.insert(0);
    c.insert(2, 4);
    c.insert(3, 6);
    c.insert(4, 1);
    c.insert(5, 1);
    c.insert(6, 1);
    c.insert(7, 1);
    int p50 = c.quantile(0.5);
    ASSERT_EQ(3, p50);
}

TEST(qdigest, hundred_int) {
    qdigest c(100);
    for (int i = 1; i <= 100; ++i) {
        c.insert(i);
    }
    int p95 = c.quantile(0.95);
    ASSERT_EQ(p95, 95);
}

File: js/test/node/qdigest_tests.js
"use strict";

var tape = require('tape');
var factory = require('../../js.js');

tape("qdigest cf=100 quantiles", function (test) {
    factory().then((sp) => {
        var seq = [];
        for (var i = 0; i < 100000; ++i)
            seq.push(1000 * Math.random());

        var gk = new sp.QDigest(100);
        for (var i = 0; i < seq.length; ++i)
            gk.insert(seq[i]);
        test.ok(gk.quantile(0.1) >= 70 && gk.quantile(0.1) <= 130);
        test.ok(gk.quantile(0.5) >= 470 && gk.quantile(0.5) <= 530);
        test.ok(gk.quantile(0.9) >= 870 && gk.quantile(0.9) <= 930);
        test.end();
    });
});
